### PR TITLE
Add Spark Variant type infrastructure [databricks]

### PR DESCRIPTION
**Pull request by nartal1**
Contributes to #15220 and #15221. Split from #15370.

Stack 1 of 4. Review and merge in order: #15642 -> #15643 -> #15644 -> #15645.

### Description

Adds the shared Spark Variant type infrastructure needed before GPU extraction can be enabled:

- Represent Variant columns using Spark's value/metadata binary children and accept both cuDF `LIST<UINT8>` and `STRING` child layouts.
- Add Variant-aware GPU/host conversion, memory estimation, null literals, Parquet schema checks, and GPU Parquet read plumbing.
- Keep Variant writes and planner registration disabled; extraction is added and activated by the follow-up PRs.

Validation:

- Spark 4.0/Scala 2.13 reactor build: passed.
- Spark 3.5.9 compatibility build: passed.
- `GpuColumnVectorVariantSuite`: 4 passed.
- `GpuColumnarToRowSuite`: Variant cases passed with ANSI mode enabled and disabled.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 1]
```diff
@@ -101,8 +101,30 @@ public static void debug(String name, HostColumnVectorCore hostCol) {
...
+  public static boolean isVariantType(DataType spark) {
+    return spark != null && "variant".equals(spark.typeName());
+  }
+
+  private static HostColumnVector.ListType variantByteListType(boolean nullable) {
+    return new HostColumnVector.ListType(
+        nullable, new HostColumnVector.BasicType(false, DType.UINT8));
+  }
+
+  public static HostColumnVector.DataType[] variantHostChildren() {
```
Comment (firestarman) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java]: should-fix: Could these two physical-layout helpers stay out of the public API? `variantHostChildren` is only used by the same package, and `variantHostType` is only used inside this class. Making the first package-private and the second private would avoid exposing the Variant cuDF STRUCT/LIST representation as a downstream API.

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/RapidsHostColumnVectorCore.java [hunk 2]
```diff
@@ -199,11 +202,16 @@ public final byte[] getBinary(int rowId) {
   @Override
   public final ColumnVector getChild(int ordinal) {
     if (cachedChildren[ordinal] == null) {
-      StructType st = (StructType) dataType();
-      StructField[] fields = st.fields();
-      for (int i = 0; i < fields.length; i++) {
-        HostColumnVectorCore tmp = cudfCv.getChildColumnView(i);
-        cachedChildren[i] = new RapidsHostColumnVectorCore(fields[i].dataType(), tmp);
+      if (GpuColumnVector.isVariantType(dataType())) {
```
Comment (firestarman) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/RapidsHostColumnVectorCore.java:205]: should-fix: Could we add an end-to-end host Variant test for this child-materialization branch? Existing tests cover a standalone STRING/BinaryType or only the device-layout predicate, so a STRUCT wrapped as VariantType and `getChild(0/1).getBinary(0)` would directly protect this path.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBatchUtils.scala [hunk 3]
```diff
@@ -88,6 +88,12 @@ object GpuBatchUtils {
         dt.fields.map { f =>
           minGpuMemory(f.dataType, f.nullable, rowCount, includeOffset)
         }.sum
+      case dt if GpuColumnVector.isVariantType(dt) =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBatchUtils.scala:91]: should-fix: Could the Spark 4 Variant suite cover the new minimum-size, estimated-size, and variable-width behavior? In particular, assert both offset buffers, the payload estimate, nullable validity overhead, `includeOffset = false`, and that Variant is not fixed-width.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala [hunk 4]
```diff
@@ -236,6 +236,8 @@ object GpuScalar extends Logging {
    */
   def from(v: Any, t: DataType): Scalar = t match {
     case nullType if v == null => nullType match {
+      case variantType if GpuColumnVector.isVariantType(variantType) =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala:239]: should-fix: Could we add a regression test for `GpuScalar(null, VariantType)`? It should force scalar construction and assert an invalid `DType.STRUCT`, protecting this new null Variant branch and its child schema.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala [hunk 5]
```diff
@@ -818,6 +820,21 @@ protected case class GpuParquetFileFilterHandler(
     }
   }
 
+  private def isVariantPhysicalType(fileType: Type): Boolean = {
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala]: should-fix: Could this physical-schema predicate get focused positive and rejection tests? The cases should cover a valid group plus primitive input, wrong child count or name, an optional child, a nested child, and a non-BINARY child so each acceptance condition and incompatibility path is exercised.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala [hunk 6]
```diff
@@ -818,6 +820,21 @@ protected case class GpuParquetFileFilterHandler(
...
+    if (fileType.isPrimitive || fileType.asGroupType().getFieldCount != 2) {
+      false
+    } else {
+      val groupType = fileType.asGroupType()
+      Seq("value", "metadata").zipWithIndex.forall { case (name, index) =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala]: must-fix: This positional check rejects a valid Parquet Variant group whose required BINARY fields appear as `metadata, value`. Spark 4 resolves the children by name and exposes them canonically as `value, metadata`; the GPU path should likewise validate by name and reorder the clipped group before cuDF materialization, otherwise a CPU-readable file can fail on the GPU path.

File: sql-plugin/src/main/spark400/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+   */
+  def toRapidsOrNull(t: DataType): DType = {
+    t match {
+      case _: DayTimeIntervalType =>
+        // use int64 as Spark does
+        DType.INT64
+      case _: YearMonthIntervalType =>
+        // use int32 as Spark does
+        DType.INT32
+      case dt if GpuColumnVector.isVariantType(dt) =>
```
Comment (firestarman) [inline sql-plugin/src/main/spark400/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala]: should-fix: Could we retain a single common `GpuTypeShims` implementation? The Spark 4 Variant case here is unreachable because `GpuColumnVector.getNonNestedRapidsType` already returns `DType.STRUCT` before the shim is consulted. Splitting the full file across Spark 3.3 and 4.0 duplicates broad policy code without version-dependent behavior.

**Review by firestarman — COMMENTED** [review]
Inline review comments.

**Comment by nartal1** [comment]
build

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 8]
```diff
@@ -101,8 +101,30 @@ public static void debug(String name, HostColumnVectorCore hostCol) {
     TableDebug.get().debug(name, hostCol);
   }
 
+  public static boolean isVariantType(DataType spark) {
```
Comment (jihoonson) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:104]: nit


```suggestion
  public static boolean isVariantType(DataType sparkType) {
```

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 9]
```diff
@@ -101,8 +101,30 @@ public static void debug(String name, HostColumnVectorCore hostCol) {
...
+    return spark != null && "variant".equals(spark.typeName());
```
Comment (jihoonson) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:105]: Can the `spark` ever be null? It looks strange that null is checked here but not in the callers such as `convertFrom()`.

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java [hunk 10]
```diff
@@ -723,6 +756,28 @@ static boolean typeConversionAllowed(ColumnView cv, DataType colType) {
...
   }
 
+  private static boolean isVariantColumn(ColumnView cv) {
+    if (!cv.getType().equals(DType.STRUCT) || cv.getNumChildren() != 2) {
+      return false;
+    }
+    try (ColumnView value = cv.getChildColumnView(0);
+         ColumnView metadata = cv.getChildColumnView(1)) {
+      return isVariantBinaryChild(value) && isVariantBinaryChild(metadata);
+    }
```
Comment (jihoonson) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:766]: So do we assume it is a variant column if the column view is a struct with two children of either strings or uint8 lists? Can there be false positives? If so, will it be a problem?

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/RapidsHostColumnVectorCore.java (hunk 2 above)
Comment (jihoonson) [inline sql-plugin/src/main/java/com/nvidia/spark/rapids/RapidsHostColumnVectorCore.java:205]: Why do we need to specialize the variant type here? Please add a comment to explain it.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala [hunk 11]
```diff
@@ -533,6 +533,8 @@ protected case class GpuParquetFileFilterHandler(
           schemaBuilder.addChild(field.name, convertToParquetNative(field.dataType))
         }
         schemaBuilder.build()
+      case dt if GpuColumnVector.isVariantType(dt) =>
+        new ParquetFooter.ValueElement()
```
Comment (jihoonson) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala:537]: Why do we return the ValueElement?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala [hunk 12]
```diff
@@ -209,10 +209,28 @@ object GpuTypeShims {
    */
   def additionalCsvSupportedTypes: TypeSig = TypeSig.DAYTIME
 
+  /**
+   * Get additional types supported for both Parquet reads and writes for this Shim
+   */
+  def additionalParquetReadWriteSupportedTypes: TypeSig = TypeSig.ansiIntervals
```
Comment (jihoonson) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala:215]: It looks quite confusing as we now have `additionalParquetSupportedTypes` and `additionalParquetReadWriteSupportedTypes`. It is not quite clear what's difference between these two until you read their code definition. I'm not sure what can be a better name than this though.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBatchUtils.scala [hunk 13]
```diff
@@ -122,6 +128,8 @@ object GpuBatchUtils {
         dt.fields.map { f =>
           estimateGpuMemory(f.dataType, f.nullable, rowCount)
         }.sum
+      case dt if GpuColumnVector.isVariantType(dt) =>
+        calculateOffsetBufferSize(rowCount) * 2 + dt.defaultSize * rowCount
```
Comment (jihoonson) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBatchUtils.scala:132]: [The default size of the variant type is 2k](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/types/VariantType.scala#L31-L33), which seems pretty large. Have we measured the impact of this over estimation?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala (hunk 4 above)
Comment (jihoonson) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala:239]: Why do we support constructing a scalar only from nulls, but not from non-nulls?

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java (hunk 9 above)
Comment (jihoonson) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:105]: And why do we check the name here? Instead of checking the type.

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java (hunk 8 above)
Comment (nartal1) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java]: Done

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java (hunk 10 above)
Comment (nartal1) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java:772]: `hasVariantPhysicalLayout` is only called after the requested Spark type has been identified as Variant, so an ordinary Spark STRUCT cannot be mistaken for Variant. I renamed the helper and added a comment to make that distinction clear

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/RapidsHostColumnVectorCore.java (hunk 2 above)
Comment (nartal1) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/RapidsHostColumnVectorCore.java:207]: Added a comment. Spark models Variant as atomic, while cuDF stores it as a STRUCT containing value and metadata. The specialization exposes those physical children through Spark’s binary accessors. Tests cover both STRING and LIST<UINT8> child representations

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala (hunk 11 above)
Comment (nartal1) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/parquet/GpuParquetScan.scala]: Thanks for pointing this out. Changed this path. Schemas containing Variant now bypass native footer pruning and use the Java footer reader, preserving the complete Variant physical group. The conversion path throws if it is reached unexpectedly

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala (hunk 12 above)
Comment (nartal1) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/shims/GpuTypeShims.scala]: Renamed the shared set to `additionalParquetCommonSupportedTypes`. The read and write specific methods now derive from that common set

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBatchUtils.scala (hunk 13 above)
Comment (nartal1) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBatchUtils.scala:134]: I have not measured this estimate separately. `GpuBatchUtils` estimates output memory from the schema and row count before Parquet data is decoded, so the actual Variant payload size is unavailable at that point. This reuses Spark’s 2048 byte defaultSize estimate and separately accounts for the two cuDF offset buffers. Spark notes that 2048 was chosen arbitrarily, so I think  tuning it with representative Variant data can be a follow-up. [Spark source](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/types/VariantType.scala#L31-L34)

File: sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java (hunk 9 above)
Comment (nartal1) [reply sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java]: Good point. None of the callers pass a null DataType, so I removed the explicit null check. The typeName comparison was originally used because the shared code also compiles against pre Spark-4 versions where VariantType is unavailable. I replaced it with version-specific shims: Spark 4.x compares directly with VariantType where as  pre Spark-4 shims return false.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala (hunk 4 above)
Comment (nartal1) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala:241]: Spark 4 stores a non-null Variant literal as a VariantVal containing the encoded value and metadata. cudf-spark does not yet convert non-null Variant literals into cuDF’s Variant STRUCT representation, so they remain on CPU. This PR supports typed null Variant scalars, which are needed to create null Variant columns. I added an integration test verifying that a non-null Variant literal falls back to CPU. Non-Null variant literal will be added as a follow on - https://github.com/NVIDIA/cudf-spark/issues/15171

**Comment by nartal1** [comment]
build

**Review by jihoonson — APPROVED** [review]

**Review by firestarman — APPROVED** [review]
gltm, only some nits, not blockers

File: integration_tests/src/main/python/asserts.py [hunk 14]
```diff
@@ -565,7 +565,8 @@ def do_it_all(spark):
 
 def assert_gpu_fallback_collect(func,
         cpu_fallback_class_name,
-        conf={}):
+        conf={},
```
Comment (firestarman) [inline integration_tests/src/main/python/asserts.py:568]: nit: Please use `conf=None` and initialize it inside the function to avoid a mutable default.

File: integration_tests/src/main/python/asserts.py [hunk 15]
```diff
@@ -584,6 +585,8 @@ def assert_gpu_fallback_collect(func,
         gpu_end - gpu_start, cpu_end - cpu_start))
     if should_sort_locally():
         _sort_locally(from_cpu, from_gpu)
+    if result_canonicalize_func_before_compare is not None:
```
Comment (firestarman) [inline integration_tests/src/main/python/asserts.py:588]: nit: Please canonicalize before `_sort_locally`; raw `VariantVal` values are not orderable and can fail before the callback runs.

File: sql-plugin/src/test/spark400/scala/com/nvidia/spark/rapids/GpuColumnVectorVariantSuite.scala [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+    withResource(ColumnVector.fromStrings("value")) { value =>
+      withResource(ColumnVector.fromStrings("metadata")) { metadata =>
+        assertHostVariantChildren(value, metadata, valueBytes, metadataBytes)
+      }
+    }
+
+    withResource(ColumnVector.fromLists(byteListType, boxedBytes(valueBytes))) { value =>
+      withResource(ColumnVector.fromLists(
+          byteListType, boxedBytes(metadataBytes))) { metadata =>
+        assertHostVariantChildren(value, metadata, valueBytes, metadataBytes)
```
Comment (firestarman) [inline sql-plugin/src/test/spark400/scala/com/nvidia/spark/rapids/GpuColumnVectorVariantSuite.scala:151]: nit: Could we directly test `convertFrom(VariantType, ...)` and `from(Array(VariantType))` to verify the generated value/metadata byte-list schema?
